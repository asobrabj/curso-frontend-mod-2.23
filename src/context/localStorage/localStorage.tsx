import { createContext, useContext, useEffect, useState } from 'react'
import { IFormData } from '../../components/FormAddAnnotation'
import IList from '../../types/lists'

export interface ILocalStorageContext {
  anotations: Array<IFormData> | null
  setAnotations: (data: IFormData | Array<IFormData>) => void
  getFilters: (value: string) => void
  getTags: (key: 'category' | 'priority', value: string) => void
  getStorage: (key: string) => Array<IFormData | IList> | string | null
  setStorage: (key: string, data: Array<IFormData | IList> | string) => void
}

interface ILocalStorageProvider {
  children: React.ReactNode
}

const LocalStorageContext = createContext<ILocalStorageContext>(
  {} as ILocalStorageContext
)

export function LocalStorageProvider({ children }: ILocalStorageProvider) {
  const [anotations, setAnotations] = useState<Array<IFormData> | null>([])

  useEffect(() => {
    const anotations = getStorage('anotations')
    setAnotations(anotations)
  }, [])

  const getTags = (key: 'category' | 'priority', value: string) => {
    const anotations = getStorage('anotations')
    const data = anotations?.filter(
      (item: IFormData) =>
        item[key]?.toLowerCase().includes(value.toLowerCase())
    )

    if (data) setAnotations(data)
    return
  }

  const getFilters = (value: string) => {
    const anotations = getStorage('anotations')
    const name = anotations?.filter((item: IFormData) =>
      item.name.toLowerCase().includes(value.toLowerCase())
    )

    if (name) setAnotations(name)
    return
  }

  const setAnotationsFn = (data: IFormData | Array<IFormData>) => {
    const storage = getStorage('anotations')
    let newStorage: Array<IFormData> = []
    if (storage) {
      newStorage = [...storage, data]
    }

    if (data instanceof Array) {
      newStorage = data
    }

    setAnotations(newStorage)
    localStorage.setItem('anotations', JSON.stringify(newStorage))
  }

  const getStorage = (key: string) => {
    const storedData = localStorage.getItem(key)
    if (storedData) {
      return JSON.parse(storedData)
    }
    return []
  }

  const saveToStorage = (
    key: string,
    data: Array<IFormData | IList> | string
  ) => {
    localStorage.setItem(key, JSON.stringify(data))
    return
  }
  return (
    <LocalStorageContext.Provider
      value={{
        getTags,
        getFilters,
        anotations,
        setAnotations: setAnotationsFn,
        getStorage,
        setStorage: saveToStorage,
      }}
    >
      {children}
    </LocalStorageContext.Provider>
  )
}

export function useStorage() {
  return useContext(LocalStorageContext)
}
