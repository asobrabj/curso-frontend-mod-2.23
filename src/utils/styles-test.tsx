import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
