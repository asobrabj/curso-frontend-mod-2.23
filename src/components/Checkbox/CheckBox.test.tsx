// import { fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import CheckBox from '.'
// import { screen } from '@testing-library/react'
import { fireEvent } from '@testing-library/react'
import * as localStorageModule from '../../context/localStorage/localStorage'
import { renderTheme } from '../../utils/styles-test'

jest.mock('../../context/localStorage/localStorage', () => ({
  useStorage: jest.fn(() => ({
    getStorage: jest.fn(() => []),
    setAnotations: jest.fn(() => []),
  })),
}))

describe('CheckBox Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('deve renderizar corretamente', () => {
    const { container } = renderTheme(<CheckBox id={1} checked={false} />)
    expect(container).toMatchSnapshot()
  })

  it('deve atualizar o estado isChecked ao clicar no checkbox', () => {
    const { getByLabelText } = renderTheme(<CheckBox id={1} checked={false} />)
    const checkbox = getByLabelText('concluido')

    fireEvent.click(checkbox)
    expect(checkbox).toBeChecked()
  })

  it('deve atualizar as anotações ao clicar no checkbox', () => {
    const mockGetStorage = jest.fn(() => [{ id: 1, concluded: false }])
    const mockSetAnotations = jest.fn()

    const spyUseStorage = jest.spyOn(localStorageModule, 'useStorage')
    spyUseStorage.mockReturnValue({
      getStorage: mockGetStorage,
      setAnotations: mockSetAnotations,
    })

    const { getByLabelText } = renderTheme(<CheckBox id={1} checked={false} />)
    const checkbox = getByLabelText('concluido')

    fireEvent.click(checkbox)

    expect(mockGetStorage).toHaveBeenCalledWith('anotations')
    expect(mockSetAnotations).toHaveBeenCalledWith([{ id: 1, concluded: true }])

    spyUseStorage.mockRestore()
  })
})
