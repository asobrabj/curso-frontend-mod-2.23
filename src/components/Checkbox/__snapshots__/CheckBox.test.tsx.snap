// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox Component deve renderizar corretamente 1`] = `
.c0 {
  display: flex;
  color: #999;
  font-weight: bold;
}

.c0 >input {
  margin-right: 8px;
}

<div>
   
  <div
    class="c0"
  >
    <input
      id="concluded"
      type="checkbox"
    />
    <label
      for="concluded"
    >
      concluido
    </label>
  </div>
   
</div>
`;
